### PR TITLE
fix: ensure keep alive is completed in time

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

ensure keep alive is completed in time

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Improved lease management and leader handling in the election process with enhanced timeout error handling.

- **Bug Fixes**
  - Adjusted default replica values for Greptimedb and Etcd clusters from 3 to 1, improving resource efficiency.

- **Refactor**
  - Switched function call from `common_runtime::spawn_bg` to `common_runtime::spawn_write` for better asynchronous operation handling.

- **Performance Improvements**
  - Extended lease duration for the metasrv leader from 3 to 5 seconds, enhancing stability.

### CODE REVIEW

File: src/meta-srv/src/election/etcd.rs [hunk 1]
```diff
@@ -297,3 +296,48 @@ impl Election for EtcdElection {
...
+                }
+            } else {
+                if self.is_leader.load(Ordering::Relaxed) {
+                    if let Err(e) = self
+                        .leader_watcher
+                        .send(LeaderChangeMessage::StepDown(Arc::new(leader.clone())))
+                    {
+                        error!(e; "Failed to send leader change message");
+                    }
+                }
```
Comment (fengjiachun) [inline src/meta-srv/src/election/etcd.rs:333]: Let's just return the error, and handle the Stepdown outside

File: src/meta-srv/src/election/etcd.rs [hunk 2]
```diff
@@ -235,39 +236,36 @@ impl Election for EtcdElection {
...
 
+            if self.is_leader.load(Ordering::Relaxed) {
+                if let Err(e) = self
+                    .leader_watcher
+                    .send(LeaderChangeMessage::StepDown(Arc::new(leader.clone())))
+                {
+                    error!(e; "Failed to send leader change message");
+                }
+            }
             self.is_leader.store(false, Ordering::Relaxed);
```
Comment (fengjiachun) [inline src/meta-srv/src/election/etcd.rs:269]:
```suggestion
            if self
                .is_leader
                .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
                .is_ok()
            {
                if let Err(e) = self
                    .leader_watcher
                    .send(LeaderChangeMessage::StepDown(Arc::new(leader.clone())))
                {
                    error!(e; "Failed to send leader change message");
                }
            }
```

**Review by fengjiachun — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]

File: src/meta-srv/src/metasrv.rs [hunk 3]
```diff
@@ -450,7 +450,7 @@ impl Metasrv {
             {
                 let election = election.clone();
                 let started = self.started.clone();
-                let _handle = common_runtime::spawn_bg(async move {
+                let _handle = common_runtime::spawn_write(async move {
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/metasrv.rs:453]: The default size of bag is too small.

File: src/meta-srv/src/election/etcd.rs [hunk 4]
```diff
@@ -231,44 +232,43 @@ impl Election for EtcdElection {
...
+                {
+                    Ok(Ok(())) => {
+                        let _ = keep_alive_interval.tick().await;
+                    }
+                    Ok(Err(err)) => {
+                        error!(err; "Failed to keep alive");
+                        break;
+                    }
+                    Err(_) => {
+                        error!("Refresh lease timeout");
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/election/etcd.rs:254]: The error must be a timeout.

File: src/meta-srv/src/election/etcd.rs [hunk 5]
```diff
@@ -231,44 +232,43 @@ impl Election for EtcdElection {
...
-                                .leader_watcher
-                                .send(LeaderChangeMessage::StepDown(Arc::new(leader.clone())))
-                            {
-                                error!("Failed to send leader change message, error: {e}");
-                            }
-                        }
+                // The keep alive operation MUST be done in `META_KEEP_ALIVE_INTERVAL_SECS`.
+                match timeout(
+                    keep_lease_duration,
+                    self.keep_alive(&mut keeper, &mut receiver, leader),
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/election/etcd.rs:242]: Yes.

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**